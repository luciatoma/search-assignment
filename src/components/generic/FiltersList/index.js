import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';
import theme from '../../../themes/default';

const availableFilters = [
    'ARTIST',
    'ARTWORK',
    'ARTICLE',
    'CITY',
    'COLLECTION',
    'FAIR',
    'FEATURE',
    'GALLERY',
    'GENE',
    'INSTITUTION',
    'PROFILE',
    'SALE',
    'SHOW',
    'TAG',
];

const StyledFilterWrapper = styled.div`
    align-items: center;
    display: flex;
    margin-top: 16px;
    position: relative;
    width: 100%;
`;

const Wrapper = styled.div`
    display: flex;
    overflow: auto;

    ::-webkit-scrollbar {
        display: none;
    }

    scrollbar-width: none;
`;

const StyledList = styled.div`
    display: flex;
    flex: 0 0 auto;
    flex-direction: row;
`;

const StyledFilter = styled.span`
    background: ${theme.color.gray1};
    border-radius: 8px;
    color: ${theme.color.black};
    cursor: pointer;
    display: flex;
    flex: 0 0 auto;
    font-family: ${theme.typeFamily.sans};
    font-size: 16px;
    letter-spacing: 1px;
    line-height: 22px;
    margin-right: 16px;
    padding: 11px 17px;
    text-transform: capitalize;

    &:last-of-type {
        margin-right: 0;
    }

    ${props =>
        props.active &&
        css`
            color: ${theme.color.white};
            font-weight: bold;
            letter-spacing: 0.4px;
            background-image: linear-gradient(0deg, ${theme.color.blue1} 3%, ${theme.color.blue2} 100%);
            text-shadow: 0 1px 0 rgba(0, 19, 25, 0.08);
        `}
`;

const ShadowSpan = styled.span`
    background: ${theme.color.white};
    height: 100%;
    position: absolute;

    ${props =>
        props.shadowLeft &&
        css`
            border-top-right-radius: 50%;
            border-bottom-right-radius: 50%;
            box-shadow: ${props.shadowLeft ? `5px 0 13px 12px ${theme.color.white}` : 'none'};
            left: 0;
            width: ${props.shadowLeft / 2}px;
        `}

    ${props =>
        props.shadowRight &&
        css`
            border-top-left-radius: 50%;
            border-bottom-left-radius: 50%;
            box-shadow: ${props.shadowRight ? `-5px 0 13px 12px ${theme.color.white}` : 'none'};
            right: 0;
            width: ${props.shadowRight / 2}px;
        `}
`;

class FiltersList extends Component {
    constructor() {
        super();

        this.state = {
            shadowLeft: 0,
            shadowRight: 0,
        };

        this.wrapper = React.createRef();
        this.list = React.createRef();
        this.activeItem = React.createRef();
    }

    componentDidMount() {
        this.handleScroll();
    }

    handleScroll = () => {
        const scrollLeft = this.wrapper && this.wrapper.current && this.wrapper.current.scrollLeft;
        const listWidth = this.list && this.list.current && this.list.current.offsetWidth;
        const wrapperWidth = this.wrapper && this.wrapper.current && this.wrapper.current.offsetWidth;
        const listScrollOutsideWrapper = wrapperWidth + scrollLeft;
        const rightScroll = listWidth - listScrollOutsideWrapper;

        if (scrollLeft < 25) this.setState({ shadowLeft: scrollLeft });

        if (listScrollOutsideWrapper >= wrapperWidth && rightScroll > 25) {
            this.setState({ shadowRight: 25 });
        } else if (listScrollOutsideWrapper >= wrapperWidth && rightScroll <= 25) {
            this.setState({ shadowRight: rightScroll });
        }

        return null;
    };

    handleFilter = (item, e) => {
        const { onFiltersChange, filters } = this.props;

        // Scroll the filter into view
        e.target.scrollIntoView({ behavior: 'smooth', block: 'nearest', inline: 'center' });
        let activeFiltersClone = [...filters];
        if (!activeFiltersClone.includes(item)) {
            activeFiltersClone.push(item);
        } else if (activeFiltersClone.includes(item)) {
            activeFiltersClone = activeFiltersClone.filter(elem => elem !== item);
        }
        this.setState({ activeFilters: activeFiltersClone });
        onFiltersChange && onFiltersChange(activeFiltersClone);
    };

    render() {
        const { shadowLeft, shadowRight } = this.state;
        const { filters } = this.props;

        return (
            <StyledFilterWrapper>
                <ShadowSpan shadowLeft={shadowLeft} />
                <Wrapper ref={this.wrapper} onScroll={this.handleScroll}>
                    <StyledList ref={this.list}>
                        {availableFilters.map(item => (
                            <StyledFilter
                                title={item.toLowerCase()}
                                active={filters && filters.includes(item)}
                                key={item}
                                onClick={e => this.handleFilter(item, e)}
                            >
                                {item.toLowerCase()}
                            </StyledFilter>
                        ))}
                    </StyledList>
                </Wrapper>
                <ShadowSpan shadowRight={shadowRight} />
            </StyledFilterWrapper>
        );
    }
}

FiltersList.propTypes = {
    onFiltersChange: PropTypes.func.isRequired,
    filters: PropTypes.array,
};

export default FiltersList;
