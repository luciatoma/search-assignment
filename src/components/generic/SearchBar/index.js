import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';
import theme from '../../../themes/default';
import magnifyingGlassIcon from '../../../images/magnifying-glass.svg';
import clearIcon from '../../../images/clear.svg';

const Wrapper = styled.div`
    max-width: 512px;
    position: relative;
    width: 100%;
`;

const StyledInput = styled.input`
    background: ${props => (props.focus ? theme.color.white : theme.color.gray1)};
    border-radius: 8px;
    box-shadow: ${props => (props.focus ? '0 1px 2px 0 rgba(0, 19, 25, 0.16), 0 4px 12px 0 rgba(0, 19, 25, 0.08)' : 'none')};
    color: ${theme.color.black};
    font-family: ${theme.typeFamily.sans};
    font-size: 18px;
    line-height: 18px;
    padding: 15px 15px 15px 52px;
    width: 100%;

    &:focus {
        outline: none;
    }
`;

const Icon = styled.img`
    cursor: pointer;
    position: absolute;

    ${props =>
        props.magnify &&
        css`
            height: 22px;
            left: 15px;
            opacity: ${props.focus ? '1' : '0.4'};
            top: 16px;
            width: 22px;
        `}

    ${props =>
        props.clear &&
        css`
            height: 18px;
            opacity: 0.4;
            right: 16px;
            top: 17px;
            width: 18px;

            &:hover {
                opacity: 1;
            }
        `}
`;

class SearchBar extends Component {
    constructor() {
        super();

        this.state = {
            focus: false,
            value: '',
        };

        this.input = React.createRef();
    }

    onFocus = () => {
        this.setState({ focus: true });
    };

    onBlur = () => {
        const { value } = this.state;
        if (!value) return this.setState({ focus: false });
        return null;
    };

    handleChange = value => {
        this.setState({ value });
    };

    handleSearch = () => {
        const { searchValue } = this.props;
        const { value } = this.state;
        searchValue && searchValue(value);
    };

    handleClear = () => {
        this.setState({ value: '' }, () => {
            this.input.current.focus();
        });
    };

    render() {
        const { focus, value } = this.state;
        console.log('value', value, focus);
        return (
            <Wrapper>
                <StyledInput
                    ref={this.input}
                    type="text"
                    placeholder="Search by artist, gallery, style, theme, tag, etc."
                    value={value}
                    focus={focus}
                    onFocus={this.onFocus}
                    onBlur={this.onBlur}
                    onChange={e => this.handleChange(e.target.value)}
                />
                <Icon magnify src={magnifyingGlassIcon} alt="Search" focus={focus} onClick={this.handleSearch} />
                {value && <Icon clear className="clearIcon" src={clearIcon} alt="Clear" onClick={this.handleClear} />}
            </Wrapper>
        );
    }
}

SearchBar.propTypes = {
    searchValue: PropTypes.func.isRequired,
};

export default SearchBar;
