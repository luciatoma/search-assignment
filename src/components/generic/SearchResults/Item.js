import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import config from '../../../config';
import theme from '../../../themes/default';
import imgPlaceholder from '../../../images/photo.svg';

const ItemWrapper = styled.a`
    align-items: center;
    display: flex;
    flex-direction: row;
    margin-top: 30px;
    position: relative;
    width: 100%;
`;

const ItemDetails = styled.div`
    display: flex;
    flex-direction: column;
    width: calc(100% - 92px);
`;

const StyledLabel = styled.span`
    color: ${theme.color.black};
    font-family: ${theme.typeFamily.sans};
    font-size: 18px;
    line-height: 24px;

    @media (min-width: 992px) {
        overflow: hidden;
        text-overflow: ellipsis;
        white-space: nowrap;
    }
`;

const StyledInfo = styled.span`
    color: ${theme.color.gray2};
    font-family: ${theme.typeFamily.sans};
    font-size: 16px;
    letter-spacing: 0;
    line-height: 22px;
    text-transform: capitalize;
`;

const ImgPlaceholder = styled.span`
    background: ${theme.color.gray1} url(${imgPlaceholder}) no-repeat center;
    border-radius: 50%;
    height: 48px;
    margin-left: auto;
    margin-right: 16px;
    width: 48px;
`;

const StyledAvatar = styled.img`
    border-radius: 50%;
    height: 48px;
    margin-left: auto;
    margin-right: 16px;
    width: 48px;
`;

const Item = props => {
    const { image, filters, label, link } = props;

    return (
        <ItemWrapper href={`${config.baseUrl}${link}`} target="_blank" title={label}>
            <ItemDetails>
                <StyledLabel>{label}</StyledLabel>
                <StyledInfo>{filters.join(', ').toLowerCase()}</StyledInfo>
            </ItemDetails>
            {image ? <StyledAvatar src={image} alt="result" /> : <ImgPlaceholder />}
        </ItemWrapper>
    );
};

Item.propTypes = {
    image: PropTypes.string,
    filters: PropTypes.array,
    label: PropTypes.string,
    link: PropTypes.string,
};

export default Item;
