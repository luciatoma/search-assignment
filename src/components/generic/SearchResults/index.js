import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import theme from '../../../themes/default';
import Item from './Item';

const Wrapper = styled.div`
    width: 100%;
`;

const EmptyResults = styled.div`
    font-family: font-family: ${theme.typeFamily.sans};
    margin-top: 30px;
`;

// const loadMore = () =>
//     fetchMore({
//         variables: {
//             cursor: comments.pageInfo.endCursor,
//         },
//         updateQuery: (previousResult, { fetchMoreResult }) => {
//             const newEdges = fetchMoreResult.comments.edges;
//             const { pageInfo } = fetchMoreResult.comments;

//             return newEdges.length
//                 ? {
//                       comments: {
//                           __typename: previousResult.comments.__typename,
//                           edges: [...previousResult.comments.edges, ...newEdges],
//                           pageInfo,
//                       },
//                   }
//                 : previousResult;
//         },
//     });

// Using index because there's nothing to uniquely identify the result in the response
/* eslint-disable react/no-array-index-key */
const SearchResults = ({ filters, results }) => {
    if (results.length === 0) {
        return <EmptyResults>No results found</EmptyResults>;
    }

    return (
        <Wrapper>
            {results.map((item, index) => {
                const result = item.node;
                return (
                    <Item key={index} label={result.displayLabel} filters={filters} image={result.imageUrl} link={result.href} />
                );
            })}
        </Wrapper>
    );
};

SearchResults.propTypes = {
    filters: PropTypes.array,
    results: PropTypes.array.isRequired,
};

export default SearchResults;
