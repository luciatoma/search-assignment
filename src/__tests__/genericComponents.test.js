import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

import SearchBar from '../components/generic/SearchBar';
import FiltersList from '../components/generic/FiltersList';
import SearchResults from '../components/generic/SearchResults';

/* global document */

// keeping the code dry
const tests = [
    { name: 'SearchBar', component: <SearchBar onQueryChange={jest.fn} /> },
    { name: 'FiltersList', component: <FiltersList onFiltersChange={jest.fn} /> },
    {
        name: 'SearchResults',
        component: (
            <SearchResults filters={['ARTIST', 'ARTWORK']} results={[{ node: { displayLabel: 'Vancouver', imageUrl: '' } }]} />
        ),
    },
];

tests.forEach(test =>
    it(`${test.name} renders without crashing and matches snapshot`, () => {
        const div = document.createElement('div');
        ReactDOM.render(test.component, div);
        ReactDOM.unmountComponentAtNode(div);
        const tree = renderer.create(test.component).toJSON();
        expect(tree).toMatchSnapshot();
    })
);
