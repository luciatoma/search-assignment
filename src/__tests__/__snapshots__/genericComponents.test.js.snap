// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FiltersList renders without crashing and matches snapshot 1`] = `
<div
  className="sc-bxivhb jGlIoF"
>
  <span
    className="sc-gzVnrw eYwnex"
  />
  <div
    className="sc-ifAKCX bATqMD"
    onScroll={[Function]}
  >
    <div
      className="sc-EHOje dxiHPF"
    >
      <span
        className="sc-bZQynM cVeDuS"
        onClick={[Function]}
        title="artist"
      >
        artist
      </span>
      <span
        className="sc-bZQynM cVeDuS"
        onClick={[Function]}
        title="artwork"
      >
        artwork
      </span>
      <span
        className="sc-bZQynM cVeDuS"
        onClick={[Function]}
        title="article"
      >
        article
      </span>
      <span
        className="sc-bZQynM cVeDuS"
        onClick={[Function]}
        title="city"
      >
        city
      </span>
      <span
        className="sc-bZQynM cVeDuS"
        onClick={[Function]}
        title="collection"
      >
        collection
      </span>
      <span
        className="sc-bZQynM cVeDuS"
        onClick={[Function]}
        title="fair"
      >
        fair
      </span>
      <span
        className="sc-bZQynM cVeDuS"
        onClick={[Function]}
        title="feature"
      >
        feature
      </span>
      <span
        className="sc-bZQynM cVeDuS"
        onClick={[Function]}
        title="gallery"
      >
        gallery
      </span>
      <span
        className="sc-bZQynM cVeDuS"
        onClick={[Function]}
        title="gene"
      >
        gene
      </span>
      <span
        className="sc-bZQynM cVeDuS"
        onClick={[Function]}
        title="institution"
      >
        institution
      </span>
      <span
        className="sc-bZQynM cVeDuS"
        onClick={[Function]}
        title="profile"
      >
        profile
      </span>
      <span
        className="sc-bZQynM cVeDuS"
        onClick={[Function]}
        title="sale"
      >
        sale
      </span>
      <span
        className="sc-bZQynM cVeDuS"
        onClick={[Function]}
        title="show"
      >
        show
      </span>
      <span
        className="sc-bZQynM cVeDuS"
        onClick={[Function]}
        title="tag"
      >
        tag
      </span>
    </div>
  </div>
  <span
    className="sc-gzVnrw jrvbvw"
  />
</div>
`;

exports[`SearchBar renders without crashing and matches snapshot 1`] = `
<div
  className="sc-bdVaJa cpqOBv"
>
  <input
    className="sc-bwzfXH erGLxu"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    placeholder="Search..."
    type="text"
    value=""
  />
  <img
    alt="Search"
    className="sc-htpNat NRooi"
    onClick={[Function]}
    src="magnifying-glass.svg"
  />
  
</div>
`;

exports[`SearchResults renders without crashing and matches snapshot 1`] = `
<div
  className="sc-jTzLTM esfPQV"
>
  <a
    className="sc-htoDjs fFONZT"
    href="https://www.artsy.netundefined"
    target="_blank"
    title="Vancouver"
  >
    <div
      className="sc-dnqmqq bRLzkm"
    >
      <span
        className="sc-iwsKbI eDvzxh"
      >
        Vancouver
      </span>
      <span
        className="sc-gZMcBi MqHje"
      >
        artist, artwork
      </span>
    </div>
    <span
      className="sc-gqjmRU bJcmDL"
    />
  </a>
</div>
`;
